### ISSUE TITLE
loadData() modifies the input parameter configdbJson, which is surprise to the caller

### ISSUE THREAD

**Issue by qiluo-msft** [Azure/sonic-buildimage#8808]
Reason for failure is the sonic-yang-mgmt pkg we consume modifies an input parameter in-place as a side-effect, I don't think this should happen since the function should only be loading the data into sonic_yang.

Here is how:
https://github.com/Azure/sonic-buildimage/blob/e6699a0132c141b000af99521eab1fabf6942921/src/sonic-yang-mgmt/sonic_yang_ext.py#L702
```python
    """
    load_data: load Config DB, crop, xlate and create data tree from it. (Public)
    input:    data
              debug Flag
    returns:  True - success   False - failed
    """
    def loadData(self, configdbJson, debug=False):

       try:
          # write Translated config in file if debug enabled
          xlateFile = None
          if debug:
              xlateFile = "xlateConfig.json"
          self.jIn = configdbJson                              ## <=== in this line self.jIn shares the same reference as configdbJson
          # reset xlate and tablesWithOutYang
          self.xlateJson = dict()
          self.tablesWithOutYang = dict()
          # self.jIn will be cropped
          self._cropConfigDB()                                ## <=== in this line self.jIn is modified and so does configdbJson
```

check the comments starting with `## <===` above

In the `config apply-patch` code we rely on the loadData function, so when we pass a configdb json object sometimes we end up with it getting modified as a side-effect.

_Originally posted by @ghooo in https://github.com/Azure/sonic-utilities/issues/1830#issuecomment-922187055_

**Comment by qiluo-msft** [comment]
@praveen-li Could you help check and triage this issue? I believe this is a bug.

**Comment by praveen-li** [comment]
@qiluo-msft  

The comment says "# self.jIn will be cropped"
If YANG models are not present for any table, then that config will be cropped.

**Comment by qiluo-msft** [comment]
I understood you need crop the `self.jIn` internally.

My concern is `configdbJson` is user input, and you crop it at the same time because both are just references. Is it a surprise to caller that a "load" function modified the "data"? As a workaround, caller must backup the data before calling `loadData()`

**Comment by ghooo** [comment]
@praveen-li This is the issue about the cropping of input parameters.

Assume we have yang models for: TABLE_A, TABLE_B
Assume configDbJson input to loadData to be:
```
{
  'TABLE_A': {},
  'TABLE_B': {},
  'TABLE_C': {}
}
```

The data will be loaded into `SonicYang` object. But at the same time there will be a side-effect of changing the value in the input configDbJson object

Full example:
```
configDbJson = {
  'TABLE_A': {},
  'TABLE_B': {},
  'TABLE_C': {}
}
print(configDbJson)
# output: { 'TABLE_A': {}, 'TABLE_B': {}, 'TABLE_C': {} }
sy = sonic_yang.SonicYang(YANG_DIR)
sy.loadYangModel()
sy.loadData(configDbJson)

print(configDbJson)
# output: { 'TABLE_A': {}, 'TABLE_B': {} }     <=== see how TABLE_C was removed from input json object.
```

**Comment by praveen-li** [comment]
Yes, this in done intentional. Because only Table A and Table B can be loaded in Data tree. Table C is removed since it can not be validated. Keeping TAble C gives the impression that it is validated, which is incorrect. This way User of library knows the non validated part if a copy is made before by user.

A copy can be made internal in library, but the issue is, how user will know that some part of the config is not validated. Please let me know your suggestions.

Thanks,

Regards
Praveen
________________________________

**Comment by qiluo-msft** [comment]
Is it possible to return a read-only reference to the internal copy? so user will know what is actually loaded. If anything failed, originally this function will return False, and we can let it return an empty tree since nothing is loaded.
